### ISSUE TITLE
Discovery of remote projects

### ISSUE THREAD

**Issue by kim** [radicle-dev/radicle-link#58]
## What

When the set of projects-of-interest is not known, and the local peer is not a long-running process with decent connectivity which just listens on the gossip, there needs to be a way to enumerate the projects a remote peer has. For this to be meaningful, it can't be just the project id, but should include (a subset of) the metadata.

## Alternatives

1. Define a dedicated RPC

    i.e. `ListProjects -> [Project]`

2. Require a well-known project to be maintained by every peer

    The project would have either an english name (`index.html`), or just the peer id. The contents would be some blobs containing an index of the projects replicated by the peer, along with metadata, and perhaps some kind of "activity feed" of recent updates (although that could be implicit via the history).

3. A combination

  A listing of minimal project info can be implemented rather efficiently (`git-web` uses the `description` file), and keeping around the updates is probably only interesting for long-lived / frequently frequented nodes (i.e. _seeds_).

**Comment by massimiliano-mantione** [comment]
The problem with solution 2 is twofold:
- the "index" blob would need to be maintained, and
- it would need to be transferred entirely every time it is updated.

Every attempt to make this solution better makes it more and more similar to approach 1.

About solution 1, the main problem here is filtering.

I mean, just transferring project IDs is not enough, because the receiver would have no way to decide if any of those projects is interesting for them.
They would have to transfer also the metadata associated to each unknown project.
And then, likely, they would have to implement some kind of search locally, over the set of transferred metadata, so that the result is meaningful for a human.

Trying to make this work implies building a sort of remote query system.
This implies several parts, including:

- a RPC protocol that can express queries
- an engine that can interpret them and execute them on the remote side
- a combination of caching and indexing to manage the execution of those searches (the naive implementation "open the metadata file for each project, deserialize it, and run the filter on it" does not scale at all)

I am seriously considering using graphql as RPC protocol (remember that it is transport agnostic).
And using sqlite as a storage layer for the "device master repo" instead of git.

This has a couple of downsides, but the advantages would be several:

* clear protocol definition, both on the network and for the application
* application uses a protocol for which they already have tooling
* application (and users) can query projects very easily
* the engineering of the storage layer would be way better that whatever we would invent otherwise
* same for the query system
* much of the implementation can be specified declaratively
* the storage layer would be "typed", removing a lot of ambiguities about file parsing and error handling

The two disadvantages I see are the following:

* git implements a form of "sync updates" natively (fetch a set of branches) which we would lose: the graphql schema would have to expose a similar query, which is not complex at all
* we would likely need to implement a git backing store for "TUF managed scoped entities", if and when we want to support them actively

If this idea is not shot sown on the spot we should likely discuss it a bit elsewhere to evaluate the tradeoffs better :-)

**Comment by kim** [comment]
Thanks, @massimiliano-mantione. I think I understand now how we ended up here. 

There is a fundamental problem with your proposal: no response from any GraphQL endpoint on the network carries more truthiness than any other. All the shenanigans with merkle-DAGs, TUF rules, refsigs, etc. etc. are there to make it possible to detect if someone is lying. Implementing a generalised query language against some storage in a decentralised system is -- basically a blockchain (the state store responds with the query result + a merkle proof).

So, I think the confusion (also on my part) which lead to this ticket is: the `radicle-link` protocol does not at all need to know anything about "soft" metadata, and specifically does not need to provide a distributed search engine for users to query metadata keywords. All it does is to replicate repositories according to a set of rules, require _some_ form of verification which lets us talk about integrity and _intended_ purpose of the repository, and lastly to gossip updates amongst interested parties.

Everything else is an application, or, to be more on-brand, let's call them _applets_.

One such applet is issue tracking, another one attaching code to issues. The most general form of this is the SSB-style "feed". Recall that we are borrowing the SSB "social overlay" paradigm: data is replicated along the edges of a social graph. So, to enter the network, one needs some weak form of trust: _interest_ in a specific project. 

Now, what's a project? Actually, anything one has a local applet for. If the entry point for a user is a "pub" server, we can mandate that such servers _SHOULD_ provide a feed (ie. a project) of things they are currently serving, and can make the datastructures conveyed in this feed as rich as we want. We can also provide one implementation of such server software, which exposes a GraphQL API to be used in a more traditional client-server fashion: after all, those nodes are supposed to have stable addresses and connectivity, and the trust model is very simple (the same as any other website).

Distributed _and_ decentralised search is an interesting problem, too, but I think its main hardness lies in incentivisation. I also think we can postpone it: we have `radicle-registry` as a global directory, and we have queryable nodes providing a local view.

It is possible that I'm missing or misunderstanding something about this proposal, but at the moment I'm inclined to close this ticket in favor of a task to define a semantics for a "Node-local Updates Feed" applet in the same spirit as #53, #57

**Comment by massimiliano-mantione** [comment]
> It is possible that I'm missing or misunderstanding something about this proposal

From your comment it seems we are not trying to address the same problem.

Specifically, the "graphql schema" idea does not try to solve the "distributed search" problem.
It only tries to solve the "query one single device" problem.

The "information propagation and distribution" should be built on top of that, but as a distinct layer.

My main worry about missing a remote query API for devices is that if we don't have one the only possible way to "explore" a device is to clone it fully, and then search the local clone.

This is not the way we are used to work.

On github (and other "centralized project hubs" we have on the web) I can browse and search the "project collection", and then decide which projects I am interested in and clone them.

Or when interacting with an issue tracker I am usually not forced to clone the whole issue tracker (including its history) just to understand if there's already an issue (eventually closed) similar to the one I want to open.

In a similar way we should probably allow users to only store the "issue feeds" they are participating to instead of the whole issue tracker, if the only think they want to do is to report a bug and discuss about it.

So, to reiterate: this has nothing to do with "distributed search".
It could be a building block to implement it in the future, but it is not addressing _that_ problem now.

This is simply that I feel that we need a "device query API" in order to be able to browse a single device to decide what we are interested in among the contents that it carries.

The most basic form of "device query" is "list everything that you have, with enough info so that I understand what it is and if it is relevant to me".
I think this is not enough for engineering reasons.

To sum it up: 

> at the moment I'm inclined to close this ticket in favor of a task to define a semantics for a "Node-local Updates Feed" applet

This would work, as long as we address the problem of searching among those feeds before subscribing to them.
Because that is the problem I was talking about.
But maybe, for reasons I don't see, this is a non-issue.

**Comment by kim** [comment]
> But maybe, for reasons I don't see, this is a non-issue.

The metadata per-project is a few bytes, maybe kilobytes at most.

**Comment by FintanH** [comment]
I think some clarity might shed some light on this conversation. When reading through it the question of: "What is the metadata?", came up for me. Without answering this the GraphQL and SQLite store doesn't seem to make any sense, to me at least.

From my point of view, it seems like the metadata that's needed are things like:
* Project ID
* Maintainers
* Description
* Possibly something like tags

Something I see as useful is peers storing this information regardless of whether they want to follow the project or not. This is so that peers of peers can view this metadata and decided if _they_ are interested. The act of being interested would then be cloning the project itself.

This is what I'm understanding from the conversation so far, but it could be possible that I'm missing the point. Let me know if something I said doesn't make sense or has engineering issues.

**Comment by kim** [comment]
@FintanH It is the metadata as defined in the spec document -- although most projects won't have all fields.

**Comment by FintanH** [comment]
Whoops 😳 I'll take a look at the spec tomorrow and revise my response 😁

**Comment by kim** [comment]
Superseded by #84 

Let me know if this needs further discussion -- I don't think there is an issue for the initial generation of the network, we can handle onboarding and discovery in a variety of ways. 

Furthermore, I don't think there is a viable solution for online queries over the global search space until there exists a stable network of mostly-available nodes (and that's not just the two or three we run). Even then, the proposed solution will need to answer what the incentives are to participate: since we can't query _all_ nodes, queries must be routed to those who are "responsible" for a search term. Traditionally, this is done using a DHT (it's not full-text search), and the economics of filesharing networks incentivise peers to advertise new data on the network. This does not directly translate to code collaboration.

**tl;dr** Product, economics, and adoption considerations will heavily influence if and how we would drive this forward.
